CONTRIBUTING.md

```diff
@@ -0,0 +1,303 @@
+```
+
+As for Nickel, you can make the `nls` binary available in your path via `cargo
+install`:
+
+```shell
+$ cargo install --path .
+$ nls --version
+nickel-lang-lsp 0.1.0
+```
+
+### WebAssembly REPL
+
+This repository is able to build WebAssembly (WASM) version of the REPL, which
+is used for the online playground on [nickel-lang.org][nickel-lang.org]. While
+using `cargo` directly is the recommended way for building, both methods have
+advantages for the WASM REPL. Building through Cargo alone is possible, but
+requires installing new tools and patching the `Cargo.toml`. On the other hand,
+`Nix` can perform the build in one simple command, but incremental compilation
+is not as good as with direct usage of `cargo`.
+
+Both methods are described below.
+
+#### Using Nix
+
+At the root of the repository, run:
+
+```shell
+$ nix build .#buildWasm
+```
+
+The corresponding NPM package is put in `result/`.
+
+#### Using Cargo
+
+1. [Install `wasm-pack`][install-wasm-pack]
+2. Add the following line to `Cargo.toml` under the `[lib]` heading:
+
+   ```diff
+   [lib]
+   +crate-type = ["cdylib", "rlib"]
+   ```
+
+3. Remove the following line from `Cargo.toml` (dependency on
+   `nickel-lang-utilities`):
+
+   ```diff
+   -nickel-lang-utilities = {path = "utilities", version = "0.1.0"}
+   ```
+
+4. Run `wasm-pack`:
+
+   ```shell
+   wasm-pack build -- --no-default-features --features repl-wasm
+   ```
+
+   A `pkg` directory, containing the corresponding NPM package, should now be
+   available.
+5. (Optional) the generated NPM package is named `nickel`, but this name is not
+   very descriptive, and is already in use in the NPM registry. You can patch
+   the name of the NPM package using `jq` to be `nickel-repl` instead:
+
+   ```shell
+   $ jq '.name = "nickel-repl"' pkg/package.json > package.json.patched \
+     && rm -f pkg/package.json \
+     && mv package.json.patched pkg/package.json
+   ```
+
+   If you don't have `jq`, you can do it by hand: replace the `name` attribute
+   in `pkg/package.json` by `"nickel-repl"`.
+
+The `pkg` directory now contains a `nickel-repl` NPM package that can be used
+from JavaScript.
+
+## Testing
+
+Tests are run via `cargo test`. They are two types of tests:
+
+- Unit tests, located directly in the corresponding module.
+- Integration tests, located in the dedicated crate `tests`.
+
+You can take inspiration from the existing tests to add your own. By convention,
+tests expected to pass are written in a standalone Nickel file in `tests/pass/`.
+Each `.ncl` file defines a list of expressions that must individually evaluate
+to the boolean `true`. The whole file is an expression that returns true if and
+only if every tests pass, or fail with a contract failure to help locating the
+failing test (instead of returning just `false`).
+
+If a test expected to pass is failing, run it directly using nickel with `nickel
+-f tests/pass/test_that_doesnt_pass.ncl` to get better error messages than
+`cargo test`.
```

Comment:
Is that a bug or misfeature in `cargo test`? Is it something that we could considering fixing upstream?

Reply:
I don't think that's really the job of `cargo` (which is lower-level). It's just that tests are usually based on `assert!` or the like, and when those fail, we get a message that two Rust objects (usually printed in a verbose and exhaustive way) are not equal. In that case, probably that `Ok(true)` is not equal to `Err(...some contract error)`.

On the other hand, running the nickel interpreter will pretty print the error as usual, pointing the right locations and so. I imagine this is technically possible to make the error reporting of tests to be more similar to the interpreter. I just don't know if it's really worth the effort: failing tests is a rather rare event, and in that case running `nickel` once separately for debugging purpose doesn't sound like a prohibitive cost.

Reply:
I see. I understand how this isn't a most urgent situation. It kind of depends on the cost to fix. It's better to have informative error messages right in the test framework, but it's also not so much better as to justify spending a lot of time on it.